File: dotnet/Rezel/Grammars/HtmlGrammar.cs
using Rezel.Common;
using Rezel.Lr;
using Rezel.Highlight;

namespace Rezel.Grammars;

public static class HtmlGrammar
{
    private class ElementContext
    {
        public string Name;
        public ElementContext? Parent;

        public ElementContext(string name, ElementContext? parent)
        {
            Name = name;
            Parent = parent;
        }
    }

    private static readonly HashSet<string> SelfClosers = [
        "area", "base", "br", "col", "command", "embed", "frame", "hr",
        "img", "input", "keygen", "link", "meta", "param", "source",
        "track", "wbr", "menuitem",
    ];

    private static readonly HashSet<string> ImplicitlyClosed = [
        "dd", "li", "optgroup", "option", "p", "rp", "rt",
        "tbody", "td", "tfoot", "th", "tr",
    ];

    private static readonly Dictionary<string, HashSet<string>> CloseOnOpen = new()
    {
        ["dd"] = ["dd", "dt"],
        ["dt"] = ["dd", "dt"],
        ["li"] = ["li"],
        ["option"] = ["option", "optgroup"],
        ["optgroup"] = ["optgroup"],
        ["p"] = [
            "address", "article", "aside", "blockquote", "dir", "div", "dl",
            "fieldset", "footer", "form", "h1", "h2", "h3", "h4", "h5", "h6",
            "header", "hgroup", "hr", "menu", "nav", "ol", "p", "pre",
            "section", "table", "ul",
        ],
        ["rp"] = ["rp", "rt"],
        ["rt"] = ["rp", "rt"],
        ["tbody"] = ["tbody", "tfoot"],
        ["td"] = ["td", "th"],
        ["tfoot"] = ["tbody"],
        ["th"] = ["td", "th"],
        ["thead"] = ["tbody", "tfoot"],
        ["tr"] = ["tr"],
    };

    private const int LessThan = 60;
    private const int GreaterThan = 62;
    private const int Slash = 47;
    private const int Question = 63;
    private const int Bang = 33;
    private const int Dash = 45;

    private static bool HtmlNameChar(int ch)
    {
        return ch == 45 || ch == 46 || ch == 58 || (ch >= 65 && ch <= 90) || ch == 95 || (ch >= 97 && ch <= 122) || ch >= 161;
    }

    private static string? TagNameAfter(InputStream input, int offset)
    {
        var next = input.Peek(offset);
        var name = "";
        while (true)
        {
            if (!HtmlNameChar(next)) break;
            name += (char)next;
            offset++;
            next = input.Peek(offset);
        }
        if (name.Length > 0) return name.ToLowerInvariant();
        return next == Question || next == Bang ? null : "";
    }

    private static bool InForeignElement(object? context)
    {
        var cx = context as ElementContext;
        while (cx != null)
        {
            if (cx.Name == "svg" || cx.Name == "math") return true;
            cx = cx.Parent;
        }
        return false;
    }

    private static string TermName(Stack stack, int term)
    {
        return stack.Parser.TermNames?.GetValueOrDefault(term) ?? "";
    }

    private static int? DialectIndex(LRParser parser, string name)
    {
        var keys = parser.Dialects.Keys.ToArray();
        var idx = Array.IndexOf(keys, name);
        return idx >= 0 ? idx : null;
    }

    private static ITokenizer MakeContentTokenizer(string tag, int textToken, int endToken)
    {
        var tagChars = tag.Select(c => (int)c).ToArray();
        var lastState = 2 + tagChars.Length;

        return new ExternalTokenizer((input, stack) =>
        {
            var state = 0;
            var matchedLen = 0;
            var i = 0;
            while (true)
            {
                if (input.Next < 0)
                {
                    if (i > 0) input.AcceptToken(textToken);
                    break;
                }
                if ((state == 0 && input.Next == LessThan) ||
                    (state == 1 && input.Next == Slash) ||
                    (state >= 2 && state < lastState && input.Next == tagChars[state - 2]))
                {
                    state++;
                    matchedLen++;
                }
                else if (state == lastState && input.Next == GreaterThan)
                {
                    if (i > matchedLen)
                        input.AcceptToken(textToken, endOffset: -matchedLen);
                    else
                        input.AcceptToken(endToken, endOffset: -(matchedLen - 2));
                    break;
                }
                else if ((input.Next == 10 || input.Next == 13) && i > 0)
                {
                    input.AcceptToken(textToken, endOffset: 1);
                    break;
                }
                else
                {
                    state = 0;
                    matchedLen = 0;
                }
                input.Advance();
                i++;
            }
        });
    }

    private static ITokenizer MakeHtmlExternalTokenizer(string name, Dictionary<string, int> terms)
    {
        if (name == "tagStart")
        {
            var StartTag = terms["StartTag"];
            var StartSelfClosingTag = terms["StartSelfClosingTag"];
            var StartScriptTag = terms["StartScriptTag"];
            var StartStyleTag = terms["StartStyleTag"];
            var StartTextareaTag = terms["StartTextareaTag"];
            var StartCloseTag = terms["StartCloseTag"];
            var NoMatchStartCloseTag = terms["NoMatchStartCloseTag"];
            var MismatchedStartCloseTag = terms["MismatchedStartCloseTag"];
            var missingCloseTag = terms["missingCloseTag"];
            var IncompleteTag = terms["IncompleteTag"];
            var IncompleteCloseTag = terms["IncompleteCloseTag"];

            return new ExternalTokenizer((input, stack) =>
            {
                if (input.Next != LessThan)
                {
                    if (input.Next < 0 && stack.Context != null)
                        input.AcceptToken(missingCloseTag);
                    return;
                }
                input.Advance();
                var close = input.Next == Slash;
                if (close) input.Advance();
                var tagName = TagNameAfter(input, 0);
                if (tagName == null) return;
                if (tagName.Length == 0)
                {
                    input.AcceptToken(close ? IncompleteCloseTag : IncompleteTag);
                    return;
                }

                var parentCtx = stack.Context as ElementContext;
                var parent = parentCtx?.Name;
                if (close)
                {
                    if (tagName == parent)
                    {
                        input.AcceptToken(StartCloseTag);
                    }
                    else if (parent != null && ImplicitlyClosed.Contains(parent))
                    {
                        input.AcceptToken(missingCloseTag, endOffset: -2);
                    }
                    else
                    {
                        var noMatchIdx = DialectIndex(stack.Parser, "noMatch");
                        if (noMatchIdx.HasValue && stack.DialectEnabled(noMatchIdx.Value))
                        {
                            input.AcceptToken(NoMatchStartCloseTag);
                        }
                        else
                        {
                            var cx = stack.Context as ElementContext;
                            while (cx != null)
                            {
                                if (cx.Name == tagName) return;
                                cx = cx.Parent;
                            }
                            input.AcceptToken(MismatchedStartCloseTag);
                        }
                    }
                }
                else
                {
                    if (tagName == "script")
                        input.AcceptToken(StartScriptTag);
                    else if (tagName == "style")
                        input.AcceptToken(StartStyleTag);
                    else if (tagName == "textarea")
                        input.AcceptToken(StartTextareaTag);
                    else if (SelfClosers.Contains(tagName))
                        input.AcceptToken(StartSelfClosingTag);
                    else if (parent != null && CloseOnOpen.TryGetValue(parent, out var closeSet) && closeSet.Contains(tagName))
                        input.AcceptToken(missingCloseTag, endOffset: -1);
                    else
                        input.AcceptToken(StartTag);
                }
            }, contextual: true);
        }

        if (name == "endTag")
        {
            var EndTag = terms["EndTag"];
            var SelfClosingEndTag = terms["SelfClosingEndTag"];

            return new ExternalTokenizer((input, stack) =>
            {
                if (input.Next == Slash && input.Peek(1) == GreaterThan)
                {
                    bool selfClosing;
                    var idx = DialectIndex(stack.Parser, "selfClosing");
                    if (idx.HasValue && stack.DialectEnabled(idx.Value))
                        selfClosing = true;
                    else
                        selfClosing = InForeignElement(stack.Context);
                    input.AcceptToken(selfClosing ? SelfClosingEndTag : EndTag, endOffset: 2);
                }
                else if (input.Next == GreaterThan)
                {
                    input.AcceptToken(EndTag, endOffset: 1);
                }
            });
        }

        if (name == "commentContent")
        {
            var commentContent = terms["commentContent"];

            return new ExternalTokenizer((input, _) =>
            {
                var dashes = 0;
                for (var i = 0; i < int.MaxValue; i++)
                {
                    if (input.Next < 0)
                    {
                        if (i > 0) input.AcceptToken(commentContent);
                        break;
                    }
                    if (input.Next == Dash)
                    {
                        dashes++;
                    }
                    else if (input.Next == GreaterThan && dashes >= 2)
                    {
                        if (i >= 3) input.AcceptToken(commentContent, endOffset: -2);
                        break;
                    }
                    else
                    {
                        dashes = 0;
                    }
                    input.Advance();
                }
            });
        }

        if (name == "scriptTokens")
        {
            var scriptText = terms["scriptText"];
            var StartCloseScriptTag = terms["StartCloseScriptTag"];
            return MakeContentTokenizer("script", scriptText, StartCloseScriptTag);
        }

        if (name == "styleTokens")
        {
            var styleText = terms["styleText"];
            var StartCloseStyleTag = terms["StartCloseStyleTag"];
            return MakeContentTokenizer("style", styleText, StartCloseStyleTag);
        }

        if (name == "textareaTokens")
        {
            var textareaText = terms["textareaText"];
            var StartCloseTextareaTag = terms["StartCloseTextareaTag"];
            return MakeContentTokenizer("textarea", textareaText, StartCloseTextareaTag);
        }

        throw new ArgumentException($"Unknown HTML external tokenizer: {name}");
    }

    private static readonly ContextTracker HtmlElementContext = new(
        start: null,
        shift: (context, term, stack, input) =>
        {
            var name = TermName(stack, term);
            if (name is "StartTag" or "StartSelfClosingTag" or "StartScriptTag" or "StartStyleTag" or "StartTextareaTag")
            {
                var tagName = TagNameAfter(input, 1) ?? "";
                return new ElementContext(tagName, context as ElementContext);
            }
            return context;
        },
        reduce: (context, term, stack, _) =>
        {
            if (TermName(stack, term) == "Element" && context is ElementContext ctx)
                return ctx.Parent;
            return context;
        },
        reuse: (context, node, stack, input) =>
        {
            var typeName = stack.Parser.TermNames?.GetValueOrDefault(node.Type.Id) ?? "";
            if (typeName is "StartTag" or "OpenTag")
            {
                var tagName = TagNameAfter(input, 1) ?? "";
                return new ElementContext(tagName, context as ElementContext);
            }
            return context;
        },
        hash: (context) =>
        {
            if (context is not ElementContext ctx) return 0;
            var h = 0;
            var c = ctx;
            while (c != null)
            {
                h = HashCode.Combine(h, c.Name);
                c = c.Parent;
            }
            return h;
        },
        @strict: false
    );

    private static readonly NodePropSource HtmlHighlighting = HighlightUtil.StyleTags(new Dictionary<string, object>
    {
        ["Text RawText IncompleteTag IncompleteCloseTag"] = Tags.Content,
        ["StartTag StartCloseTag SelfClosingEndTag EndTag"] = Tags.Bracket,
        ["TagName"] = Tags.TypeName,
        ["MismatchedCloseTag/TagName"] = new[] { Tags.TypeName, Tags.Invalid },
        ["AttributeName"] = Tags.PropertyName,
        ["AttributeValue UnquotedAttributeValue"] = Tags.AttributeValue,
        ["Is"] = Tags.DefinitionKeyword,
        ["EntityReference CharacterReference"] = Tags.Character,
        ["Comment"] = Tags.Comment,
        ["ProcessingInst"] = Tags.ProcessingInstruction,
        ["DoctypeDecl"] = Tags.DocumentMeta,
    });

    private static LRParser? _parser;
    public static LRParser Parser => _parser ??= CreateParser();

    private static LRParser CreateParser()
    {
        var termTable = HtmlParserData.TermTable!;
        var externals = new Dictionary<string, ITokenizer>
        {
            ["scriptTokens"] = MakeHtmlExternalTokenizer("scriptTokens", termTable),
            ["styleTokens"] = MakeHtmlExternalTokenizer("styleTokens", termTable),
            ["textareaTokens"] = MakeHtmlExternalTokenizer("textareaTokens", termTable),
            ["endTag"] = MakeHtmlExternalTokenizer("endTag", termTable),
            ["tagStart"] = MakeHtmlExternalTokenizer("tagStart", termTable),
            ["commentContent"] = MakeHtmlExternalTokenizer("commentContent", termTable),
        };
        var spec = HtmlParserData.MakeSpec(
            externals: externals,
            propSources: [HtmlHighlighting],
            context: HtmlElementContext
        );
        return new LRParser(spec);
    }
}

File: swift/Sources/Rezel/Grammars/HtmlGrammar.swift
import Foundation

// MARK: - Element Context

private class ElementContext {
	let name: String
	let parent: ElementContext?

	init(name: String, parent: ElementContext?) {
		self.name = name
		self.parent = parent
	}
}

// MARK: - HTML Data Tables

private let selfClosers: Set<String> = [
	"area", "base", "br", "col", "command", "embed", "frame", "hr",
	"img", "input", "keygen", "link", "meta", "param", "source",
	"track", "wbr", "menuitem",
]

private let implicitlyClosed: Set<String> = [
	"dd", "li", "optgroup", "option", "p", "rp", "rt",
	"tbody", "td", "tfoot", "th", "tr",
]

private let closeOnOpen: [String: Set<String>] = [
	"dd": ["dd", "dt"],
	"dt": ["dd", "dt"],
	"li": ["li"],
	"option": ["option", "optgroup"],
	"optgroup": ["optgroup"],
	"p": [
		"address", "article", "aside", "blockquote", "dir", "div", "dl",
		"fieldset", "footer", "form", "h1", "h2", "h3", "h4", "h5", "h6",
		"header", "hgroup", "hr", "menu", "nav", "ol", "p", "pre",
		"section", "table", "ul",
	],
	"rp": ["rp", "rt"],
	"rt": ["rp", "rt"],
	"tbody": ["tbody", "tfoot"],
	"td": ["td", "th"],
	"tfoot": ["tbody"],
	"th": ["td", "th"],
	"thead": ["tbody", "tfoot"],
	"tr": ["tr"],
]

private let lessThan = 60
private let greaterThan = 62
private let slash = 47
private let question = 63
private let bang = 33
private let dash = 45

// MARK: - Helpers

private func htmlNameChar(_ ch: Int) -> Bool {
	return ch == 45 || ch == 46 || ch == 58 || (ch >= 65 && ch <= 90) || ch == 95 || (ch >= 97 && ch <= 122) || ch >= 161
}

private func tagNameAfter(_ input: InputStream, _ offset: Int) -> String? {
	var offset = offset
	var next = input.peek(offset)
	var name = ""
	while true {
		if !htmlNameChar(next) { break }
		name += String(UnicodeScalar(next)!)
		offset += 1
		next = input.peek(offset)
	}
	if !name.isEmpty {
		return name.lowercased()
	}
	return next == question || next == bang ? nil : ""
}

private func inForeignElement(_ context: ElementContext?) -> Bool {
	var cx = context
	while let c = cx {
		if c.name == "svg" || c.name == "math" { return true }
		cx = c.parent
	}
	return false
}

private func termName(_ stack: Stack, _ term: Int) -> String {
	return stack.parser.termNames?[term] ?? ""
}

private func dialectIndex(_ parser: LRParser, _ name: String) -> Int? {
	let keys = Array(parser.dialects.keys)
	return keys.firstIndex(of: name)
}

// MARK: - External Tokenizer Factory

private func makeHtmlExternalTokenizer(name: String, terms: [String: Int]) -> TokenizerProtocol {
	if name == "tagStart" {
		let StartTag = terms["StartTag"]!
		let StartSelfClosingTag = terms["StartSelfClosingTag"]!
		let StartScriptTag = terms["StartScriptTag"]!
		let StartStyleTag = terms["StartStyleTag"]!
		let StartTextareaTag = terms["StartTextareaTag"]!
		let StartCloseTag = terms["StartCloseTag"]!
		let NoMatchStartCloseTag = terms["NoMatchStartCloseTag"]!
		let MismatchedStartCloseTag = terms["MismatchedStartCloseTag"]!
		let missingCloseTag = terms["missingCloseTag"]!
		let IncompleteTag = terms["IncompleteTag"]!
		let IncompleteCloseTag = terms["IncompleteCloseTag"]!

		return ExternalTokenizer({ input, stack in
			if input.next != lessThan {
				if input.next < 0, stack.context != nil {
					input.acceptToken(missingCloseTag)
				}
				return
			}
			input.advance()
			let close = input.next == slash
			if close { input.advance() }
			guard let name = tagNameAfter(input, 0) else { return }
			if name.isEmpty {
				input.acceptToken(close ? IncompleteCloseTag : IncompleteTag)
				return
			}

			let parent = (stack.context as? ElementContext)?.name
			if close {
				if name == parent {
					input.acceptToken(StartCloseTag)
				} else if let parent = parent, implicitlyClosed.contains(parent) {
					input.acceptToken(missingCloseTag, endOffset: -2)
				} else if let noMatchIdx = dialectIndex(stack.parser, "noMatch"),
				          stack.dialectEnabled(noMatchIdx)
				{
					input.acceptToken(NoMatchStartCloseTag)
				} else {
					var cx = stack.context as? ElementContext
					while let c = cx {
						if c.name == name { return }
						cx = c.parent
					}
					input.acceptToken(MismatchedStartCloseTag)
				}
			} else {
				if name == "script" {
					input.acceptToken(StartScriptTag)
				} else if name == "style" {
					input.acceptToken(StartStyleTag)
				} else if name == "textarea" {
					input.acceptToken(StartTextareaTag)
				} else if selfClosers.contains(name) {
					input.acceptToken(StartSelfClosingTag)
				} else if let parent = parent,
				          let closeSet = closeOnOpen[parent],
				          closeSet.contains(name)
				{
					input.acceptToken(missingCloseTag, endOffset: -1)
				} else {
					input.acceptToken(StartTag)
				}
			}
		}, contextual: true)
	}

	if name == "endTag" {
		let EndTag = terms["EndTag"]!
		let SelfClosingEndTag = terms["SelfClosingEndTag"]!

		return ExternalTokenizer { input, stack in
			if input.next == slash, input.peek(1) == greaterThan {
				let selfClosing: Bool
				if let idx = dialectIndex(stack.parser, "selfClosing"), stack.dialectEnabled(idx) {
					selfClosing = true
				} else {
					selfClosing = inForeignElement(stack.context as? ElementContext)
				}
				input.acceptToken(selfClosing ? SelfClosingEndTag : EndTag, endOffset: 2)
			} else if input.next == greaterThan {
				input.acceptToken(EndTag, endOffset: 1)
			}
		}
	}

	if name == "commentContent" {
		let commentContent = terms["commentContent"]!

		return ExternalTokenizer { input, _ in
			var dashes = 0
			for i in 0 ..< Int.max {
				if input.next < 0 {
					if i > 0 { input.acceptToken(commentContent) }
					break
				}
				if input.next == dash {
					dashes += 1
				} else if input.next == greaterThan, dashes >= 2 {
					if i >= 3 {
						input.acceptToken(commentContent, endOffset: -2)
					}
					break
				} else {
					dashes = 0
				}
				input.advance()
			}
		}
	}

	if name == "scriptTokens" {
		let scriptText = terms["scriptText"]!
		let StartCloseScriptTag = terms["StartCloseScriptTag"]!
		return makeContentTokenizer("script", textToken: scriptText, endToken: StartCloseScriptTag)
	}

	if name == "styleTokens" {
		let styleText = terms["styleText"]!
		let StartCloseStyleTag = terms["StartCloseStyleTag"]!
		return makeContentTokenizer("style", textToken: styleText, endToken: StartCloseStyleTag)
	}

	if name == "textareaTokens" {
		let textareaText = terms["textareaText"]!
		let StartCloseTextareaTag = terms["StartCloseTextareaTag"]!
		return makeContentTokenizer("textarea", textToken: textareaText, endToken: StartCloseTextareaTag)
	}

	fatalError("Unknown HTML external tokenizer: \(name)")
}

private func makeContentTokenizer(_ tag: String, textToken: Int, endToken: Int) -> TokenizerProtocol {
	let tagChars = Array(tag.unicodeScalars.map { Int($0.value) })
	let lastState = 2 + tagChars.count

	return ExternalTokenizer { input, _ in
		var state = 0
		var matchedLen = 0
		var i = 0
		while true {
			if input.next < 0 {
				if i > 0 { input.acceptToken(textToken) }
				break
			}
			if (state == 0 && input.next == lessThan) ||
				(state == 1 && input.next == slash) ||
				(state >= 2 && state < lastState && input.next == tagChars[state - 2])
			{
				state += 1
				matchedLen += 1
			} else if state == lastState, input.next == greaterThan {
				if i > matchedLen {
					input.acceptToken(textToken, endOffset: -matchedLen)
				} else {
					input.acceptToken(endToken, endOffset: -(matchedLen - 2))
				}
				break
			} else if input.next == 10 || input.next == 13, i > 0 {
				input.acceptToken(textToken, endOffset: 1)
				break
			} else {
				state = 0
				matchedLen = 0
			}
			input.advance()
			i += 1
		}
	}
}

// MARK: - Context Tracker

private nonisolated(unsafe) let htmlElementContext = ContextTracker(
	start: Any?.none as Any,
	shift: { context, term, stack, input in
		let name = termName(stack, term)
		if name == "StartTag" || name == "StartSelfClosingTag" || name == "StartScriptTag" || name == "StartStyleTag" || name == "StartTextareaTag" {
			let tagName = tagNameAfter(input, 1) ?? ""
			return ElementContext(name: tagName, parent: context as? ElementContext) as Any
		}
		return context
	},
	reduce: { context, term, stack, _ in
		if termName(stack, term) == "Element", let ctx = context as? ElementContext {
			return ctx.parent as Any
		}
		return context
	},
	reuse: { context, node, stack, input in
		let type = node.type.id
		let name = stack.parser.termNames?[type] ?? ""
		if name == "StartTag" || name == "OpenTag" {
			let tagName = tagNameAfter(input, 1) ?? ""
			return ElementContext(name: tagName, parent: context as? ElementContext) as Any
		}
		return context
	},
	hash: { context in
		guard let ctx = context as? ElementContext else { return 0 }
		var h = 0
		var c: ElementContext? = ctx
		while let cur = c {
			h = h &* 11 &+ cur.name.hashValue
			c = cur.parent
		}
		return h
	},
	strict: false
)

// MARK: - Highlighting

private nonisolated(unsafe) let htmlHighlighting = styleTags([
	"Text RawText IncompleteTag IncompleteCloseTag": hlContent,
	"StartTag StartCloseTag SelfClosingEndTag EndTag": hlBracket,
	"TagName": hlTypeName,
	"MismatchedCloseTag/TagName": [hlTypeName, (hlTags["invalid"] as! Tag)],
	"AttributeName": hlPropertyName,
	"AttributeValue UnquotedAttributeValue": (hlTags["attributeValue"] as! Tag),
	"Is": (hlTags["definitionKeyword"] as! Tag),
	"EntityReference CharacterReference": (hlTags["character"] as! Tag),
	"Comment": hlComment,
	"ProcessingInst": (hlTags["processingInstruction"] as! Tag),
	"DoctypeDecl": (hlTags["documentMeta"] as! Tag),
])

// MARK: - Build Parser

public let htmlParser: LRParser = {
	let externals: [String: TokenizerProtocol] = [
		"scriptTokens": makeHtmlExternalTokenizer(name: "scriptTokens", terms: HtmlParserData.termTable),
		"styleTokens": makeHtmlExternalTokenizer(name: "styleTokens", terms: HtmlParserData.termTable),
		"textareaTokens": makeHtmlExternalTokenizer(name: "textareaTokens", terms: HtmlParserData.termTable),
		"endTag": makeHtmlExternalTokenizer(name: "endTag", terms: HtmlParserData.termTable),
		"tagStart": makeHtmlExternalTokenizer(name: "tagStart", terms: HtmlParserData.termTable),
		"commentContent": makeHtmlExternalTokenizer(name: "commentContent", terms: HtmlParserData.termTable),
	]
	let spec = HtmlParserData.makeSpec(
		externals: externals,
		propSources: [htmlHighlighting],
		context: htmlElementContext
	)
	return LRParser(spec: spec)
}()
